party_time/party_time.lua

```diff
@@ -0,0 +1,346 @@
+end)
+options = settings.load(defaults)
+options.blacklist = sets(options.blacklist)
+options.whitelist = sets(options.whitelist)
+
+local invite_dialog = {}
+local invite_dialog_state = {
+    title = 'Party Invite',
+    style = 'normal',
+    x = options.ui.x,
+    y = options.ui.y,
+    width = 179,
+    height = 96,
+    resizable = false,
+    moveable = true,
+    closable = true,
+}
+local unhandled_requests = {}
+local unhandled_invite = false
+
+-- Command Arg Types
+-- Create and Register types
+local arg_lookups = {
+    addremove = {
+        ['add'] = 'add',
+        ['a'] = 'add',
+        ['+'] = 'add',
+        ['remove'] = 'remove',
+        ['r'] = 'remove',
+        ['-'] = 'remove',
+    },
+    bool = {
+        ['yes'] = true,
+        ['y'] = true,
+        ['on'] = true,
+        ['no'] = false,
+        ['n'] = false,
+        ['off'] = false,
+    },
+}
+
+command.arg.register('name', '<name:string(%a+)>')
+
+command.arg.register_type('lookup_boolean', {
+    check = function(str)
+        local bool = arg_lookups.bool[str]
+        if bool ~= nil then
+            return bool
+        end
+
+        error('Expected one of \'' .. table.concat(args_lookups.bool) .. '\', received \'' .. str .. '\'.')
+    end
+})
+
+command.arg.register_type('lookup_add_remove', {
+    check = function(str)
+        local op = arg_lookups.addremove[str]
+        print(op, str)
+        if op ~= nil then
+            return op
+        end
+
+        error('Expected one of \'' .. table.concat(args_lookups.addremove) .. '\', received \'' .. str .. '\'.')
+    end
+})
+
+-- Addon Command Handlers
+-- Seven main commands additoinal sub_commands within
+local pt = command.new('pt')
+
+local invite = function(...)
+    for _, name in pairs({...}) do
+        command.input('/pcmd add ' .. name)
+        name = name:gsub('^%l', string.upper)
+        if unhandled_requests[name] then
+            unhandled_requests[name] = nil
+        end
+    end
+end
+
+pt:register('i', invite, '{name}*')
+pt:register('invite', invite, '{name}*')
+
+
+local request = function(name)
+    command.input('/prcmd add ' .. name)
+end
+
+pt:register('r', request, '{name}')
+pt:register('request', request, '{name}')
+
+
+local blacklist = function(sub_cmd, ...)
+    local names = sets({...})
+    if sub_cmd == 'add' then
+        options.blacklist = options.blacklist + names
+    elseif sub_cmd == 'remove' then
+        options.blacklist = options.blacklist + names
+    end
+    settings.save(options)
+end
+
+pt:register('b', blacklist, '<sub_cmd:lookup_add_remove> {name}*')
+pt:register('blacklist', blacklist, '<sub_cmd:lookup_add_remove> {name}*')
+
+
+local whitelist = function(sub_cmd, ...)
+    local names = sets({...})
+    print(sub_cmd)
+    if sub_cmd == 'add' then
+        options.whitelist = options.whitelist + names
+    elseif sub_cmd == 'remove' then
+        options.whitelist = options.whitelist - names
+    end
```

Comment:
Could replace the last 5 lines with:
```lua
    sets[sub_cmd == 'add' and 'union' or 'difference'](options.whitelist, names)
```

... not sure if you should, though :D I'd probably favor this more if you adjusted the lookups to result in `union` or `difference` instead of `add` and `remove` (since the user doesn't see that anyway). Then it would just be:
```lua
    sets[sub_cmd](options.whitelist, names)
```

Still not sure if you should, just throwing it out there ;D